### PR TITLE
[Merged by Bors] - Kerberos AD backend

### LINKED ISSUE #253: AD kerberos keytab backend

### PR DESCRIPTION
**Pull request by nightkr**
# Description

Fixes #253 

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Comment by nightkr** [comment]
Note before merging this: currently this disables TLS validation, need to let users supply a trust root instead.

**Comment by sbernauer** [comment]
Successfully tested this branch against a Windows Server 2022 AD and https://github.com/stackabletech/hdfs-operator/pull/334
image

**Comment by sbernauer** [comment]
I run into the following issue:
```
2023-03-28T06:46:48.462981Z  INFO stackable_krb5_provision_keytab: Creating principal principal="jn/hdfs-journalnode-default-0.hdfs-journalnode-default.kuttl-test-wired-rat.svc.cluster.local"                                                                                                                                           
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: LdapResult { result: LdapResult { rc: 68, matched: "", text: "00002071: UpdErr: DSID-03050438
, problem 6005 (ENTRY_EXISTS), data 0\n\0", refs: [], ctrls: [] } }', rust/krb5-provision-keytab/src/main.rs:259:22                                                  
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace 
```
```
k -n kuttl-test-wired-rat get secret secret-operator-ad-passwords -o yaml
data:
  dn--hdfs-datanode-default-0.hdfs-datanode-default.kuttl-test-wired-rat.svc.cluster.local--EXAMPLE.COM: RU5FUDVtJzNsXzZRZFhNYS8nIVwlRkRicS1FfHZfOGd8O0xueUlZYA==
  dn--hdfs-datanode-default.kuttl-test-wired-rat.svc.cluster.local--EXAMPLE.COM: cV1fYHohTjRbbXQ6e2A5fklDUkU2fT03JT0lTTd6cFg+Mjkldm40RQ==
  jn--hdfs-journalnode-default.kuttl-test-wired-rat.svc.cluster.local--EXAMPLE.COM: KD0ueGZbNyQ0TCl6eFdLcD1GTDFiUFEnd08ySlF5bkx5UitHbTZfNA==
  nn--hdfs-namenode-default-0.hdfs-namenode-default.kuttl-test-wired-rat.svc.cluster.local--EXAMPLE.COM: SSx2JU89UVtmLDJqO0phTl1fdjsjVmJQMkNJWil6VnQpYUpFQ05tVQ==
  nn--hdfs-namenode-default.kuttl-test-wired-rat.svc.cluster.local--EXAMPLE.COM: Nl99bnV1WEY5ZS5RemIsQWNscDddWTpFKE9MeXNXU1xsNUQmJnp7Rg==
  ```
Theese are all the avaialble JN principals
image

You can see my AD contains users from a previous testrun but *not* the user `jn/hdfs-journalnode-default-0.hdfs-journalnode-default.kuttl-test-wired-rat.svc.cluster.local` (only `jn/hdfs-journalnode-default-0.hdfs-journalnode-default.kuttl-test-clever-hog.svc.cluster.local` is already in the AD)

**Comment by sbernauer** [comment]
Maybe I'm stupid and it actually tries to create `dn/kind-control-plane`, which names clash (as of the node scope)
Because the Secret containing the cashed passwords is located in the kuttl namespace it get's deleted while the AD user still existing.
What do you suggest?
1. Remove the node scope from hdfs-op?
2. Somehow fix this (haven't thought too much yet)
 a. Force creation of new user
 b. Override pw of the user and store it in the Secret
 b. Somehow read the password and put it in the Secret (I guess this is not possible)
3. Something else
```
2023-03-28T07:08:51.977954Z  INFO stackable_krb5_provision_keytab: Creating principal principal="dn/kind-control-plane"                                              
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: LdapResult { result: LdapResult { rc: 68, matched: "", text: "00002071: UpdErr: DSID-03050438
, problem 6005 (ENTRY_EXISTS), data 0\n\0", refs: [], ctrls: [] } }', rust/krb5-provision-keytab/src/main.rs:259:22                                                  
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by nightkr** [comment]
There might have been a race condition, it can also happen if we fail to write the password back into K8s after creating the AD user...

File: rust/krb5-provision-keytab/src/main.rs [hunk 1]
```diff
@@ -109,20 +157,136 @@ fn run() -> Result<Response, Error> {
...
+                            ),
+                            ("unicodePwd".as_bytes(), [&*password_ad_encoded].into()),
+                            ("userAccountControl".as_bytes(), ["66048".as_bytes()].into()),
+                            (
+                                "userPrincipalName".as_bytes(),
+                                [princ_name.as_bytes()].into(),
+                            ),
+                            (
+                                "servicePrincipalName".as_bytes(),
+                                // FIXME: Assumes that principal request is given without realm, unparse without realm instead
```
Comment (sbernauer) [inline rust/krb5-provision-keytab/src/main.rs:246]: Marker

File: rust/krb5-provision-keytab/src/main.rs [hunk 2]
```diff
@@ -109,20 +157,136 @@ fn run() -> Result<Response, Error> {
...
+                                .unwrap()
+                        });
+                        pwd_utf16le
+                    };
+                    let user_dn = format!("CN={principal_cn},{user_distinguished_name}",);
+                    ldap.add(
+                        &user_dn,
+                        vec![
+                            ("cn".as_bytes(), [principal_cn.as_bytes()].into()),
+                            ("objectClass".as_bytes(), ["user".as_bytes()].into()),
```
Comment (sbernauer) [inline rust/krb5-provision-keytab/src/main.rs:231]: Do we need the user to be able to configure any of the settings below?

File: rust/krb5-provision-keytab/src/main.rs [hunk 3]
```diff
@@ -109,20 +157,136 @@ fn run() -> Result<Response, Error> {
...
+                    let password_ad_encoded = {
+                        let mut pwd_utf16le = Vec::new();
+                        format!("\"{password}\"").encode_utf16().for_each(|word| {
+                            WriteBytesExt::write_u16::<LittleEndian>(&mut pwd_utf16le, word)
+                                .unwrap()
+                        });
+                        pwd_utf16le
+                    };
+                    let user_dn = format!("CN={principal_cn},{user_distinguished_name}",);
+                    ldap.add(
```
Comment (sbernauer) [inline rust/krb5-provision-keytab/src/main.rs:227]: Can we please have at least a descriptive error message including which user it is trying to create, so that e.g. already existent users can be identified

File: tests/templates/kuttl/kerberos/secretclass.yaml [hunk 4]
```diff
@@ -9,7 +9,9 @@ spec:
     kerberosKeytab:
       realmName: CLUSTER.LOCAL
       kdc: krb5-kdc
-      adminServer: krb5-kdc
+      admin:
+        mit:
+          adminServer: krb5-kdc
```
Comment (sbernauer) [inline tests/templates/kuttl/kerberos/secretclass.yaml:14]: Do we want to simply call it `server` as the term `admin` is already defined above?

File: tests/test-definition.yaml [hunk 5]
```diff
@@ -1,5 +1,7 @@
 ---
 dimensions: []
 tests:
-  - name: kerberos
```
Comment (sbernauer) [inline tests/test-definition.yaml:4]: Marker to revert the change before merging

**Review by sbernauer — COMMENTED** [review]
Did walk through the code.
Please don't forget to add a breaking changelog entry

File: rust/krb5-provision-keytab/src/main.rs [hunk 6]
```diff
@@ -152,11 +316,29 @@ impl<T: std::error::Error> Display for Report<T> {
...
+    // Allow all ASCII alphanumeric characters as well as punctuation
+    // Exclude double quotes (") since they are used by the AD password update protocol...
+    let dict: Vec<char> = (1..=127)
+        .filter_map(char::from_u32)
+        .filter(|c| *c != '"' && (c.is_ascii_alphanumeric() || c.is_ascii_punctuation()))
+        .collect();
+    let pw = (0..len)
+        .map(|_| *dict.choose(&mut rng).unwrap())
+        .collect::<String>();
+    assert_eq!(pw.len(), len);
```
[the remark this answers is not in the document]
Comment (nightkr) [reply rust/krb5-provision-keytab/src/main.rs:342]: This "should" never fail, it's just intended as a failsafe to make absolutely sure that we don't break it.

**Comment by nightkr** [comment]
I think I'd like to get the CRD change into 23.4, which would make the rest of this PR non-breaking when it does land.

File: rust/krb5-provision-keytab/src/main.rs [hunk 7]
```diff
@@ -109,20 +157,136 @@ fn run() -> Result<Response, Error> {
...
+                    pw.0.clone()
+                } else {
+                    tracing::info!(
+                        principal = princ_req.name,
+                        cache_key = password_cache_key,
+                        "did not find principal in key cache, creating"
+                    );
+                    tracing::info!(principal = ?princ_req.name, "Creating principal");
+                    let principal_cn = ldap3::dn_escape(&*princ_req.name);
+                    // FIXME: AD restricts RDNs to 64 characters
```
[the remark this answers is not in the document]
Comment (nightkr) [reply rust/krb5-provision-keytab/src/main.rs:215]: We could break it into a DN component per hostname component, but that requires the user to have permission to create containers of some kind.

**Comment by sbernauer** [comment]
> I think I'd like to get the CRD change into 23.4, which would make the rest of this PR non-breaking when it does land.

+1, was thinking the same

**Comment by sbernauer** [comment]
Regarding b25995dfc4c65fec020d37d98ab40a7b6d8bc198: Do we also want so support webpki as ca-cert validation mechanism (as we e.g. do for ldap server conections). Or would say 99% of the users run their AD behind their own pki?

**Comment by nightkr** [comment]
I would assume that private PKI is by far the norm. We can always go back and enable WebPKI support later, shoudl the need arise.

File: docs/modules/secret-operator/pages/secretclass.adoc [hunk 8]
```diff
@@ -78,7 +78,19 @@ spec:
     kerberosKeytab:
       realmName: CLUSTER.LOCAL
       kdc: krb5-kdc
-      adminServer: krb5-kdc
+      admin:
```
[the remark this answers is not in the document]
Comment (nightkr) [reply docs/modules/secret-operator/pages/secretclass.adoc:81]: 9b7345662b5ce6c694885a7ce25dc3bc54f33fee

File: rust/krb5-provision-keytab/src/main.rs [hunk 9]
```diff
@@ -109,20 +157,136 @@ fn run() -> Result<Response, Error> {
...
+                user_distinguished_name,
+                schema_distinguished_name,
+            } => {
+                let kube = stackable_operator::client::create_client(None)
+                    .await
+                    .unwrap();
+                let princ_name = princ.to_string();
+                let mut password_cache = kube
+                    .get::<Secret>(
+                        password_cache_secret.name.as_deref().unwrap(),
```
[the remark this answers is not in the document]
Comment (nightkr) [reply rust/krb5-provision-keytab/src/main.rs]: Yep, definitely.. mostly covered this now, but there are some places in krb5 that need a second pass

File: rust/krb5-provision-keytab/src/main.rs (hunk 2 above)
Comment (nightkr) [reply rust/krb5-provision-keytab/src/main.rs]: I don't believe so, but we can always revisit if the need arises.

File: tests/templates/kuttl/kerberos/secretclass.yaml (hunk 4 above)
Comment (nightkr) [reply tests/templates/kuttl/kerberos/secretclass.yaml]: Resolved in #257

File: tests/test-definition.yaml (hunk 5 above)
Comment (nightkr) [reply tests/test-definition.yaml:4]: 8d28a9d6d6255eba2579116fa92abd5cbc8c633a

File: rust/krb5-provision-keytab/src/main.rs (hunk 3 above)
Comment (sbernauer) [reply rust/krb5-provision-keytab/src/main.rs]: thx!

File: docs/modules/secret-operator/pages/secretclass.adoc [hunk 10]
```diff
@@ -81,6 +81,16 @@ spec:
       admin:
         mit:
           kadminServer: krb5-kdc
+        # or...
+        activeDirectory:
+          # ldapServer must match the AD Domain Controller's FQDN or GSSAPI authn will fail
+          # You may need to set AD as your fallback DNS resolver in your Kube DNS Corefile
+          ldapServer: addc.example.com
+          passwordCacheSecret:
```
Comment (sbernauer) [inline docs/modules/secret-operator/pages/secretclass.adoc:89]: Can we please also document
```
          ldapTlsCaSecret:
            # namespace: default
            name: secret-operator-ad-ca
```
Also that we will always use ldaps and the Secret needs the key `ca.crt`

File: rust/krb5/src/lib.rs [hunk 11]
```diff
@@ -95,6 +113,21 @@ impl KrbContext {
             raw: principal,
         })
     }
+
+    /// Get the default realm configured for this context.
+    pub fn default_realm(&self) -> Result<DefaultRealm, Error> {
```
Comment (sbernauer) [inline rust/krb5/src/lib.rs:118]: WDYT of writing out the default kerberos realm into a dedicated file of the secret volume to make it easier for operators such as hdfs-op to pick it up?

**Review by sbernauer — CHANGES_REQUESTED** [review]
Thanks for working on this! Only two minor comments left

File: tests/test-definition.yaml [hunk 12]
```diff
@@ -3,3 +3,6 @@ dimensions: []
 tests:
   - name: kerberos
     dimensions: []
+  # Requires manual connection to an AD cluster
+  - name: kerberos-ad
```
Comment (sbernauer) [inline tests/test-definition.yaml:7]: Do you want to disable this test by default before merging?

Comment (nightkr) [reply tests/test-definition.yaml:7]: Yes, I think that makes sense to do.

File: rust/krb5/src/lib.rs (hunk 11 above)
Comment (nightkr) [reply rust/krb5/src/lib.rs:118]: I'd rather not add redundancy there, but the current HDFS situation isn't great either...

Perhaps another solution would be to add a utility to sble-base or something that uses libkrb5 to read out this stuff? (Either just reading the default realm, or something slightly more correct like "canonicalize this Kerberos principal?")

Comment (sbernauer) [reply rust/krb5/src/lib.rs:118]: Yeah, sounds like a more complicated but also better option

**Review by sbernauer — APPROVED** [review]
LGTM % the documentation comment

File: tests/test-definition.yaml (hunk 12 above)
Comment (nightkr) [reply tests/test-definition.yaml]: 723898c4316d7188e457920395f2d8873470c911

File: docs/modules/secret-operator/pages/secretclass.adoc (hunk 10 above)
Comment (nightkr) [reply docs/modules/secret-operator/pages/secretclass.adoc:92]: 77c0c05bffc39051d0dedcb5af108fa9243b2c7d

**Review by sbernauer — APPROVED** [review]
Thx!
